libc/stat.c

```diff
@@ -0,0 +1,987 @@
+
+	close(fifo_fd);
+	remove(tempPath);
+
+	TEST_ASSERT_EQUAL_INT(0, mkfifo(tempPath, 0000));
+
+	TEST_ASSERT_EQUAL_INT(0, stat(tempPath, &buffer));
+	TEST_ASSERT_TRUE(S_ISFIFO(buffer.st_mode));
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(tempPath, &buffer));
+	TEST_ASSERT_TRUE(S_ISFIFO(buffer.st_mode));
+
+	close(fifo_fd);
+	remove(tempPath);
+#endif
+}
+
+
+TEST(stat_mode, chr_type)
+{
+/* Check only on Phoenix-RTOS */
+#ifdef __phoenix__
+	struct stat buffer;
+	const char *filePath = "/dev/console";
+
+	TEST_ASSERT_EQUAL_INT(0, stat(filePath, &buffer));
+	TEST_ASSERT_TRUE(S_ISCHR(buffer.st_mode));
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(filePath, &buffer));
+	TEST_ASSERT_TRUE(S_ISCHR(buffer.st_mode));
+#endif
+}
+
+
+TEST(stat_mode, sock_type)
+{
+	int sfd;
+	struct stat buffer;
+	struct sockaddr_un addr;
+	const char *socketPath = "/tmp/my_socket";
+
+	unlink(socketPath);
+
+	/* Create a new socket. */
+	sfd = socket(AF_UNIX, SOCK_STREAM, 0);
+
+	/* Set the address for the socket. */
+	memset(&addr, 0, sizeof(struct sockaddr_un));
+	addr.sun_family = AF_UNIX;
+	strncpy(addr.sun_path, socketPath, sizeof(addr.sun_path) - 1);
+
+	/* Bind the socket to the address. */
+	bind(sfd, (struct sockaddr *)&addr, sizeof(struct sockaddr_un));
+
+/* Disabled on Phoenix-RTOS because recognize not as socket but character special file */
+#ifndef __phoenix__
+	TEST_ASSERT_EQUAL_INT(0, stat(socketPath, &buffer));
+	TEST_ASSERT_TRUE(S_ISSOCK(buffer.st_mode));
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(socketPath, &buffer));
+	TEST_ASSERT_TRUE(S_ISSOCK(buffer.st_mode));
+#endif
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(sfd, &buffer));
+	TEST_ASSERT_TRUE(S_ISSOCK(buffer.st_mode));
+
+	close(sfd);
+	unlink(socketPath);
+}
+
+////////////////////////////////
+
+TEST_SETUP(stat_nlink_size_blk_tim)
+{
+	fd = open(path, O_CREAT, 0666);
+}
+
+
+TEST_TEAR_DOWN(stat_nlink_size_blk_tim)
+{
+	close(fd);
+	remove(path);
+}
+
+
+TEST(stat_nlink_size_blk_tim, nlink)
+{
+	struct stat buffer;
+	const char *anotherPath = "test_stat_another_link_path";
+
+	TEST_ASSERT_EQUAL_INT(0, stat(path, &buffer));
+	TEST_ASSERT_EQUAL_INT(1, buffer.st_nlink);
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer));
+	TEST_ASSERT_EQUAL_INT(1, buffer.st_nlink);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer));
+	TEST_ASSERT_EQUAL_INT(1, buffer.st_nlink);
+
+
+	TEST_ASSERT_EQUAL_INT(0, link(path, symPath));
+	TEST_ASSERT_EQUAL_INT(0, link(path, tempPath));
+	TEST_ASSERT_EQUAL_INT(0, link(path, anotherPath));
+
+
+	TEST_ASSERT_EQUAL_INT(0, stat(path, &buffer));
+	TEST_ASSERT_EQUAL_INT(4, buffer.st_nlink);
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer));
+	TEST_ASSERT_EQUAL_INT(4, buffer.st_nlink);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer));
+	TEST_ASSERT_EQUAL_INT(4, buffer.st_nlink);
+
+	unlink(symPath);
+	unlink(tempPath);
+	unlink(anotherPath);
+}
+
+
+TEST(stat_nlink_size_blk_tim, size_blk_blocks)
+{
+/* Disabled because of #743 issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/743 */
```

Comment:
In this issue there is no information about incorrect setting size/blksize/blocks, it's also set in a wrong way on jffs2?